File: cleanup_planner/src/pioneer_wrapper.cpp
/*********************************************************************
 * Software License Agreement (BSD License)
 *
 *  Copyright (c) 2010, Georgia Institute of Technology
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of the Georgia Institute of Technology nor the names of
 *     its contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/

#include "pioneer_wrapper.h"

//
// Assignment of class constants
//
const int PioneerWrapper::GRIP_OPEN   = 1;
const int PioneerWrapper::GRIP_CLOSE  = 2;
const int PioneerWrapper::GRIP_STOP   = 3;
const int PioneerWrapper::LIFT_UP     = 4;
const int PioneerWrapper::LIFT_DOWN   = 5;
const int PioneerWrapper::LIFT_STOP   = 6;
const int PioneerWrapper::GRIP_STORE  = 7;
const int PioneerWrapper::GRIP_DEPLOY = 8;
const int PioneerWrapper::GRIP_HALT   = 15;
const int PioneerWrapper::GRIP_PRESS  = 16;
const int PioneerWrapper::LIFT_CARRY  = 17;

const unsigned int PioneerWrapper::GRIPPER_OPEN_STATE   = 1;
const unsigned int PioneerWrapper::GRIPPER_CLOSED_STATE = 2;
const unsigned int PioneerWrapper::GRIPPER_MOVING_STATE = 3;
const int PioneerWrapper::LIFT_STATIC_STATE    = 1;
const int PioneerWrapper::LIFT_MOVING_STATE    = 2;

const int PioneerWrapper::MOTORS_ON  = 4;
const int PioneerWrapper::MOTORS_OFF = 0;

//
// Constructors
//
PioneerWrapper::PioneerWrapper(ros::NodeHandle &n) :
    n_(n)
{
  gripper_state_sub_ = n_.subscribe("gripper_state", 1,
                                    &PioneerWrapper::gripperStateCallback,
                                    this);
  motor_state_sub_ = n_.subscribe("motor_state", 1,
                                  &PioneerWrapper::motorStateCallback,
                                  this);

  gripper_pub_ = n_.advertise<p2os::GripperState>("gripper_cmd",1000);
  motor_state_pub_ = n_.advertise<p2os::MotorState>("motor_state_cmd",1000);
  ptz_pub_ = n_.advertise<p2os::PTZState>("ptz_cmd",1000);
}

//
// Callback Functions
//

void PioneerWrapper::gripperStateCallback(const p2os::GripperStateConstPtr &msg)
{
  gripper_state_.grip.state = msg->grip.state;
  gripper_state_.grip.dir = msg->grip.dir;

  gripper_state_.grip.inner_beam = msg->grip.inner_beam;
  gripper_state_.grip.outer_beam = msg->grip.outer_beam;

  gripper_state_.grip.left_contact = msg->grip.left_contact;
  gripper_state_.grip.right_contact = msg->grip.right_contact;

  gripper_state_.lift.state = msg->lift.state;
  gripper_state_.lift.dir = msg->lift.dir;
  gripper_state_.lift.position = msg->lift.position;

}

void PioneerWrapper::motorStateCallback(const p2os::MotorStateConstPtr &msg)
{
  motor_state_.state = msg->state;
  if (!motorsActivated())
    activateMotors();
}

//
// Mid-level gripper behaviors
//
bool PioneerWrapper::grabObject()
{
  // Assume in line with object and open grippers
  if (! gripperMoving())
  {
    if( gripperOpen())
      closeGripper();
    else
      return true;
  }
  return false;
}

bool PioneerWrapper::releaseObject()
{
  // Assume gripper is closed and down
  if (! gripperMoving())
  {
    if( gripperClosed())
      openGripper();
    else
      return true;
  }
  return false;
}

bool PioneerWrapper::pickupObject()
{
  if (grabObject())
  {
    if (! liftMoving())
    {
      if (gripper_state_.lift.position < 1.0)
        raiseGripper();
      else
        return true;
    }
  }
  return false;
}

bool PioneerWrapper::putDownObject()
{
  if (! liftMoving())
  {
    if (gripper_state_.lift.position > 0.0)
      lowerGripper();
    else if (releaseObject())
      return true;
  }
  return false;
}


//
// Low Level Gripper Commands
//
void PioneerWrapper::deployGripper()
{
  p2os::GripperState g;
  g.grip.state = GRIP_OPEN;
  g.lift.state = LIFT_DOWN;
  gripper_pub_.publish(g);
}

void PioneerWrapper::storeGripper()
{
  p2os::GripperState g;
  g.grip.state = GRIP_CLOSE;
  g.lift.state = LIFT_UP;
  gripper_pub_.publish(g);
}

void PioneerWrapper::openGripper()
{
  p2os::GripperState g;
  g.grip.state = GRIP_OPEN;
  gripper_pub_.publish(g);
}

void PioneerWrapper::closeGripper()
{
  p2os::GripperState g;
  g.grip.state = GRIP_CLOSE;
  gripper_pub_.publish(g);
}

void PioneerWrapper::raiseGripper()
{
  p2os::GripperState g;
  g.lift.state = LIFT_UP;
  gripper_pub_.publish(g);
}

void PioneerWrapper::lowerGripper()
{
  p2os::GripperState g;
  g.lift.state = LIFT_UP;
  gripper_pub_.publish(g);
}

//
// Motor Commands
//
void PioneerWrapper::activateMotors()
{
  p2os::MotorState m;
  m.state = MOTORS_ON;
  motor_state_pub_.publish(m);
}

void PioneerWrapper::deactivateMotors()
{
  p2os::MotorState m;
  m.state = MOTORS_OFF;
  motor_state_pub_.publish(m);
}


//
// State Query Functions
//
bool PioneerWrapper::gripperMoving()
{
  return (gripper_state_.grip.state == GRIPPER_MOVING_STATE);
}

bool PioneerWrapper::gripperClosed()
{
  return (gripper_state_.grip.state == GRIPPER_CLOSED_STATE);
}

bool PioneerWrapper::gripperOpen()
{
  return (gripper_state_.grip.state == GRIPPER_OPEN_STATE);
}

bool PioneerWrapper::liftMoving()
{
  return (gripper_state_.lift.state == LIFT_MOVING_STATE);
}

bool PioneerWrapper::motorsActivated()
{
  return (motor_state_.state == MOTORS_ON);
}

File: pioneer_control/src/pioneer_control/pioneer_wrapper.py
#!/usr/bin/env python
import roslib; roslib.load_manifest('pioneer_control')
import rospy
import pioneer_command_constants as Commands
from p2os_driver.msg import GripperState, MotorState, PTZState, BatteryState
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry

class PioneerWrapper:
    def __init__(self):
        """
        Class to provide intermediate wrappers to perform standard tasks with
        the pionner.
        """
        # Setup publishers to the pioneer
        self.motor_pub = rospy.Publisher("motor_state_cmd", MotorState)
        self.vel_pub = rospy.Publisher("cmd_vel", Twist)
        self.gripper_pub = rospy.Publisher("gripper_cmd", GripperState)
        self.ptz_pub = rospy.Publisher("ptz_cmd", PTZState)

        # Setup subscribers to the pioneer information
        rospy.Subscriber("motor_state", MotorState, self.motor_state_callback)
        rospy.Subscriber("battery_state", BatteryState,
                         self.battery_state_callback)
        rospy.Subscriber("gripper_state", GripperState,
                         self.gripper_state_callback)
        rospy.Subscriber("ptz_status", PTZState, self.ptz_state_callback)
        rospy.Subscriber("pose", Odometry, self.odometry_callback)

        # Class state fields
        self.motor_state = MotorState()
        self.battery_state = BatteryState()
        self.ptz_state = PTZState()
        self.gripper_state = GripperState()
        self.odometry_pose = Odometry()

        # Query variables
        self.gripper_moving = False
        self.gripper_open = False
        self.gripper_closed = False
        self.lift_moving = False
        self.gripper_moving = False
        self.motors_activated = False

    #
    # Callback functions
    #
    def motor_state_callback(self, msg):
        """
        Callback method to store the current state of the motors
        """
        self.motor_state.state = msg.state
        self.motors_activated = (self.motor_state.state == Commands.MOTORS_ON)

        if not self.motors_activated:
            self.activate_motors()

    def battery_state_callback(self, msg):
        """
        Callback method to store the current state of the battery
        """
        self.battery_state.voltage = msg.voltage

    def gripper_state_callback(self, msg):
        """
        Callback method to store the current state of the gripper
        """
        self.gripper_state.grip.state = msg.grip.state
        self.gripper_state.grip.dir = msg.grip.dir

        self.gripper_state.grip.inner_beam = msg.grip.inner_beam
        self.gripper_state.grip.outer_beam = msg.grip.outer_beam

        self.gripper_state.grip.left_contact = msg.grip.left_contact
        self.gripper_state.grip.right_contact = msg.grip.right_contact

        self.gripper_state.lift.state = msg.lift.state
        self.gripper_state.lift.dir = msg.lift.dir
        self.gripper_state.lift.position = msg.lift.position

        # Set query values
        self.gripper_moving = False
        self.gripper_closed = False
        self.gripper_open = False

        state = self.gripper_state.grip.state
        if state == Commands.GRIPPER_OPEN_STATE:
            self.gripper_open = True
        elif state == Commands.GRIPPER_CLOSED_STATE:
            self.gripper_closed = True
        elif state == Commands.GRIPPER_MOVING_STATE:
            self.gripper_moving = True

        self.lift_moving = False
        if self.gripper_state.lift.state == Commands.LIFT_MOVING_STATE:
            self.lift_moving = True

    def ptz_state_callback(self, msg):
        """
        Callback method to store the current state of the PTZ
        """
        self.ptz_state.pan = msg.pan
        self.ptz_state.tilt = msg.tilt
        self.ptz_state.zoom = msg.zoom

    def odometry_callback(self, msg):
        """
        Callback method to store the current odometry state
        """
        self.odometry_pose.pose = msg.pose
        self.odometry_pose.twist = msg.twist

    #
    # Mid level gripper behaviors
    #
    def grab_object(self):
        if not self.gripper_moving:
            if self.gripper_open:
                self.close_gripper()
            else:
                return True
        return False

    def release_object(self):
        if not self.gripper_moving:
            if self.gripper_closed:
                self.open_gripper()
            else:
                return True
        return False

    def pickup_object(self):
        if self.grab_object():
            if not self.lift_moving:
                if self.gripper_state.lift.position < 1.0:
                    self.raise_gripper()
                else:
                    return True
        return False

    def put_down_object(self):
        if not self.lift_moving:
            if self.gripper_state.lift.position > 0.0:
                self.lower_gripper()
            elif self.release_object():
                return True
        return False

    #
    # Low level gripper behaviors
    #
    def deploy_gripper(self):
        cmd = GripperState()
        cmd.grip.state = Commands.GRIP_OPEN
        cmd.lift.state = Commands.LIFT_DOWN
        self.gripper_pub.publish(cmd)

    def store_gripper(self):
        cmd = GripperState()
        cmd.grip.state = Commands.GRIP_CLOSE
        cmd.lift.state = Commands.LIFT_UP
        self.gripper_pub.publish(cmd)

    def open_gripper(self):
        cmd = GripperState()
        cmd.grip.state = Commands.GRIP_OPEN
        self.gripper_pub.publish(cmd)

    def close_gripper(self):
        cmd = GripperState()
        cmd.grip.state = Commands.GRIP_CLOSE
        self.gripper_pub.publish(cmd)

    def raise_gripper(self):
        cmd = GripperState()
        cmd.lift.state = Commands.LIFT_UP
        self.gripper_pub.publish(cmd)

    def lower_gripper(self):
        cmd = GripperState()
        cmd.lift.state = Commands.LIFT_DOWN
        self.gripper_pub.publish(cmd)

    #
    # Motor state control
    #
    def activate_motors(self):
        m = MotorState()
        m.state = Commands.MOTORS_ON
        self.motor_pub.publish(m)

    def deactivate_motors(self):
        m = MotorState()
        m.state = Commands.MOTORS_OFF
        self.motor_pub.publish(m)
